source/core/lethegridtools.cc

```diff
@@ -0,0 +1,982 @@
+//
+// Created by lucka on 2021-07-15.
+//
+
+#include <core/lethegridtools.h>
+#include <deal.II/grid/manifold.h>
+#include <deal.II/grid/manifold_lib.h>
+#include <deal.II/grid/grid_tools.h>
+#include <deal.II/fe/fe_q.h>
+#include <cmath>
+
+
+template <int dim>
+void
+LetheGridTools::vertices_cell_mapping(const DoFHandler<dim> &dof_handler,
+        std::map<unsigned int,std::set<typename DoFHandler<dim>::active_cell_iterator>> & vertices_cell_map)
+{
+    // Find all the cells around each vertices
+
+    vertices_cell_map.clear();
+    const auto &cell_iterator = dof_handler.active_cell_iterators();
+
+
+    // // Loop on all the cells and find their vertices to fill the map of sets of
+    // cells around each vertex
+    for (const auto &cell : cell_iterator)
+    {
+        if (cell->is_locally_owned() || cell->is_ghost())
+        {
+            const unsigned int vertices_per_cell =
+                    GeometryInfo<dim>::vertices_per_cell;
+            for (unsigned int i = 0; i < vertices_per_cell; i++)
+            {
+                // First obtain vertex index
+                unsigned int v_index = cell->vertex_index(i);
+
+                // Insert the cell into the set of cell around that vertex.
+                vertices_cell_map[v_index].insert(cell);
+            }
+        }
+    }
+}
+
+
+template <int dim>
+typename DoFHandler<dim>::active_cell_iterator
+LetheGridTools::find_cell_around_point_with_tree(const DoFHandler<dim> &dof_handler,
+                                 Point<dim>             &point)
+{
+    // find cell around point using tree properties instead of looping on all cell
+
+    // define stuff for the search
+    MappingQ1<dim> map;
+    const auto &   cell_iterator = dof_handler.cell_iterators_on_level(0);
+    unsigned int   max_childs    = GeometryInfo<dim>::max_children_per_cell;
+    typename DoFHandler<dim>::cell_iterator best_cell_iter;
+
+    bool cell_0_found = false;
+
+    // loop on the cell on lvl 0 of the mesh
+    for (const auto &cell : cell_iterator)
+    {
+        try
+        {
+            const Point<dim, double> p_cell =
+                    map.transform_real_to_unit_cell(cell, point);
+
+            const double dist = GeometryInfo<dim>::distance_to_unit_cell(p_cell);
+
+            if (dist == 0)
+            {
+                // cell on lvl 0 found
+                cell_0_found   = true;
+                best_cell_iter = cell;
+            }
+        }
+        catch (const typename MappingQGeneric<dim>::ExcTransformationFailed &)
+        {}
+    }
+
+    if (cell_0_found)
```

Comment:
Why this if? What happens if cell_0_found==false? The function does nothing? This is weird. It would be better to have if (!cell_0_found) return;  than to use this large encompassing if. This hinders readability.

Reply:
Do we have exception handling in Lethe? The situation that happens here is that if you give a point that is outside the triangulation, the function is not really valid in the sense that you can't return a valid cell iterator.

Reply:
Well, we do have some exception handling through deal.II, but we don't necessarily want to continue the simulation if things broke. What I would do is have an assertThrow here that would stop the simulation. At least it is useful in debug mode.

Reply:
corrected it now throw if the point is not in the level 0 of the mesh.